### PR TITLE
Use indices when drawing Triangles

### LINKED ISSUE #2678: Triangles are using 6 vertices when they could use 4
The same data is repeated in two of the vertices. We could do this better (in a similar way to how Quads are being drawn via indices).

### PR DESCRIPTION
**Pull request by KoziLord**
Closes #2678.

Uses the ``QuadVertexBuffer<T>`` to index the Triangles (both Quads and Triangles have four vertices and six indices)

### CODE REVIEW

File: osu.Framework/Graphics/Batches/TriangleBatch.cs [hunk 1]
```diff
@@ -20,14 +20,6 @@ public TriangleBatch(int size, int maxBuffers)
...
-        protected override VertexBuffer<T> CreateVertexBuffer() => new TriangleVertexBuffer(Size, BufferUsageHint.DynamicDraw);
-
-        private class TriangleVertexBuffer : LinearVertexBuffer<T>
-        {
-            public TriangleVertexBuffer(int size, BufferUsageHint dynamicDraw)
-                : base(size * TextureGLSingle.VERTICES_PER_TRIANGLE, PrimitiveType.Triangles, dynamicDraw)
-            {
-            }
-        }
+        protected override VertexBuffer<T> CreateVertexBuffer() => new QuadVertexBuffer<T>(Size, BufferUsageHint.DynamicDraw);
```
Comment (peppy) [inline osu.Framework/Graphics/Batches/TriangleBatch.cs]: we may want a comment here mentioning why we can use a quad buffer

**Comment by smoogipoo** [comment]
You have to be careful with how the texture is mapped.

Master:
image

This branch:
image

```csharp
public class TestSceneScratch : FrameworkTestScene
{
    public TestSceneScratch()
    {
        Add(new TexturedTriangle
        {
            Anchor = Anchor.Centre,
            Origin = Anchor.Centre,
            Size = new Vector2(300, 150)
        });
    }

    private class TexturedTriangle : Triangle
    {
        [BackgroundDependencyLoader]
        private void load(TextureStore textures)
        {
            Texture = textures.Get(@"sample-texture");
        }
    }
}
```

**Comment by KoziLord** [comment]
Should the texture mapping match the original or is it alright to change it to look like a triangle-clipped quad?
image

**Comment by smoogipoo** [comment]
For now, it should match

**Comment by KoziLord** [comment]
Matching the original would require creating a fifth vertex, a copy of the top vertex with only the texture coordinates changed which would reduce the memory savings per triangle from 88 bytes to 32 bytes.
I think changing it to not match the original branch shouldn't cause issues as I don't think that triangles are actually textured anywhere in any of the framework components or in Lazer.

**Comment by KoziLord** [comment]
I do know that, however looking at the way the texture is wrapped in the original I don't think others would texture them either.

**Comment by peppy** [comment]
I tend to agree, and propose we move forward with the way you have it.

**Comment by KoziLord** [comment]
Ready for review

**Comment by smoogipoo** [comment]
If you're gonna draw it that way, then you should be able to remove the bottom-centre vertex and use only 3 indices.

**Comment by smoogipoo** [comment]
On further discussion, it may be better to keep it drawn as a quad, but batch them together with quads.

**Pull request merged**